File: sub_projs/cllm/core/bindings/c/cli/argv.c
/* argv.c — argv 掃描 → parsed_args_t（對齊 cli.cpp 解析段）。 */
#include "argv.h"
#include "flags.h"
#include "internal.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* 各動態陣列上限＝argc（旗標／位置參數再多不超過 argv 個數）。 */
static int alloc_arrays(parsed_args_t *p, int argc) {
  memset(p, 0, sizeof *p);
  p->raw_values = calloc(argc, sizeof *p->raw_values);
  p->prompt_parts = calloc(argc, sizeof *p->prompt_parts);
  p->media_specs = calloc(argc, sizeof *p->media_specs);
  p->tool_specs = calloc(argc, sizeof *p->tool_specs);
  p->modality_specs = calloc(argc, sizeof *p->modality_specs);
  return p->raw_values && p->prompt_parts && p->media_specs && p->tool_specs && p->modality_specs;
}

void parsed_free(parsed_args_t *p) {
  free(p->raw_values); free((void *)p->prompt_parts); free((void *)p->media_specs);
  free((void *)p->tool_specs); free((void *)p->modality_specs);
  memset(p, 0, sizeof *p);
}

int parse_argv(int argc, char **argv, parsed_args_t *p) {
  if (!alloc_arrays(p, argc)) { parsed_free(p); fputs("記憶體不足\n", stderr); return EXIT_USAGE; }
  int i = 1, no_more_flags = 0;
  while (i < argc) {
    char *a = argv[i];
    if (no_more_flags) { p->prompt_parts[p->prompt_n++] = a; i++; continue; }
    if (!strcmp(a, "--")) { no_more_flags = 1; i++; continue; }
    if (!strcmp(a, "--help") || !strcmp(a, "-h")) { print_usage(stderr); return EXIT_OK; }
    if (!strcmp(a, "--stream")) { p->stream = 1; i++; continue; }

    /* 需要吃下一個 argv 的旗標：缺值＝用法錯。 */
    const reflect_flag_t *rf = flag_lookup(a);
    int wants_value = rf || !strcmp(a, "--image") || !strcmp(a, "--media") ||
                      !strcmp(a, "--schema") || !strcmp(a, "--system") || !strcmp(a, "--config") ||
                      !strcmp(a, "--tool") || !strcmp(a, "--modality") || !strcmp(a, "--media-out");
    if (wants_value) {
      if (i + 1 >= argc) { fprintf(stderr, "%s 缺少值（llm-c --help 看用法）\n", a); return EXIT_USAGE; }
      char *v = argv[++i];
      if (rf) { p->raw_values[p->raw_n++] = (raw_value_t){rf->field, rf->type, rf->flag, v}; }
      else if (!strcmp(a, "--image") || !strcmp(a, "--media")) p->media_specs[p->media_n++] = v;
      else if (!strcmp(a, "--schema")) { p->schema_text = v; p->has_schema = 1; }
      else if (!strcmp(a, "--system")) { p->system_text = v; p->has_system = 1; }
      else if (!strcmp(a, "--config")) { p->config_path = v; p->has_config = 1; }
      else if (!strcmp(a, "--tool")) p->tool_specs[p->tool_n++] = v;
      else if (!strcmp(a, "--modality")) p->modality_specs[p->modality_n++] = v;
      else if (!strcmp(a, "--media-out")) p->media_out_dir = v;
      i++;
      continue;
    }
    if (a[0] == '-' && strcmp(a, "-")) { /* 「-」＝stdin 佔位符；其餘 '-' 開頭＝未知旗標 */
      fprintf(stderr, "未知旗標：%s（llm-c --help 看用法）\n", a);
      return EXIT_USAGE;
    }
    p->prompt_parts[p->prompt_n++] = a;
    i++;
  }
  return -1; /* 解析成功 */
}

File: sub_projs/handy/pllm/pllm/argv.py
"""argv.py — 命令列掃描：把 argv 拆成旗標與位置參數（對齊 cli.cpp 的解析段）。

固定旗標（--stream/--image/--schema/--system/--config/--tool/--modality/--media-out/--help/--）
特判；反射旗標（連線／取樣，對應 Client 欄位，見 flags.py）吃下一個 argv；其餘當位置參數拼
prompt。「-」單獨＝stdin 插入點，其餘 '-' 開頭＝未知旗標。掃描結果收成 ParsedArgs；遇 --help
或用法錯時回一個退出碼交呼叫端直接返回（main 只讀結果、不重演解析）。
"""
import sys

from .flags import FLAG_TO_FIELD, print_usage
from .internal import EXIT_OK, EXIT_USAGE


class ParsedArgs:
    """argv 掃描結果：旗標收成的各欄位，交給 cli.main 續組 prompt／config／請求。"""

    def __init__(self):
        self.raw_values = {}       # 反射欄位名 → (原始字串, 型別, flag)
        self.prompt_parts = []
        self.media_specs = []
        self.tool_specs = []
        self.modality_specs = []
        self.schema_text = None
        self.config_path = None
        self.media_out_dir = None
        self.system_text = None
        self.has_schema = False
        self.has_config = False
        self.has_system = False
        self.stream = False


def parse_argv(args):
    """掃描 argv，回 (ParsedArgs, None)；遇 --help／用法錯回 (None, 退出碼)。"""
    p = ParsedArgs()
    i = 1
    n = len(args)

    def need_value(flag):
        """吃下一個 argv 當旗標的值；缺值回 None（呼叫端據此回 EXIT_USAGE）。"""
        nonlocal i
        if i + 1 >= n:
            sys.stderr.write("%s 缺少值（llm --help 看用法）\n" % flag)
            return None
        i += 1
        return args[i]

    no_more_flags = False
    while i < n:
        a = args[i]
        if no_more_flags:
            p.prompt_parts.append(a)
            i += 1
            continue
        if a == "--":
            no_more_flags = True
            i += 1
            continue
        if a in ("--help", "-h"):
            print_usage()
            return None, EXIT_OK
        if a == "--stream":
            p.stream = True
            i += 1
            continue
        if a in ("--image", "--media"):
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.media_specs.append(v)
            i += 1
            continue
        if a == "--schema":
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.schema_text, p.has_schema = v, True
            i += 1
            continue
        if a == "--system":
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.system_text, p.has_system = v, True
            i += 1
            continue
        if a == "--config":
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.config_path, p.has_config = v, True
            i += 1
            continue
        if a == "--tool":
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.tool_specs.append(v)
            i += 1
            continue
        if a == "--modality":
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.modality_specs.append(v)
            i += 1
            continue
        if a == "--media-out":
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            p.media_out_dir = v
            i += 1
            continue
        if a in FLAG_TO_FIELD:
            v = need_value(a)
            if v is None:
                return None, EXIT_USAGE
            field, typ = FLAG_TO_FIELD[a]
            p.raw_values[field] = (v, typ, a)
            i += 1
            continue
        if a and a[0] == "-" and a != "-":  # 「-」＝stdin 佔位符；其餘 '-' 開頭＝未知旗標
            sys.stderr.write("未知旗標：%s（llm --help 看用法）\n" % a)
            return None, EXIT_USAGE
        p.prompt_parts.append(a)
        i += 1

    return p, None
